sycl/source/detail/scheduler/commands.cpp

```diff
@@ -308,20 +331,22 @@ cl_int ReleaseCommand::enqueueImp() {
                     : MAllocaCmd->MLinkedAllocaCmd->getMemAllocation();
 
     MemoryManager::unmap(MAllocaCmd->getSYCLMemObj(), Dst, Queue, Src,
-                         std::move(RawEvents), UnmapEvent);
+                         RawEvents, UnmapEvent);
 
     std::swap(MAllocaCmd->MIsActive, MAllocaCmd->MLinkedAllocaCmd->MIsActive);
-    RawEvents.push_back(UnmapEvent);
+    EventImpls.clear();
+    EventImplPtr UnmapEventImpl(
+        new event_impl(UnmapEvent, Queue->get_context()));
```

Comment:
Where is this deleted? Isn't this a leak?
Also should we rather create SYCL event out of PI event, and event_impl directly?

Reply:
EventImplPtr is a shared_ptr<event_impl> class object. This constructor will use the ptr returned from new as the ptr to the managed object. So as per my understanding, when the reference count becomes zero, delete is called on the ptr, and the event_impl destructor gets called.
I can check my theory with a simple c++ test code and confirm to you.

Why do you suggest creating a SYCL event object? I thought we were planning to use only impl* objects in out internal representations. Plus event_impl handles the Pi APi calls to eventretain and eventrelease..

Reply:
I checked again. The constructor that uses PiEvent is more like a copy constructor. So it calls a eventRetain and eventRelease in the desctructor. So the memory leak issue will still be there.
I changed the code to construct an event_impl directly. That will take care of the memory leak.

Reply:
Right, with shared_prt this should be fine.
And thanks for properly releasing the created PI event.

Reply:
As an idea: we can just release the unmap event right after calling MemoryManager::release

Reply:
@garimagu Could you please consider my suggestion above?

Reply:
Please, disregard my previous comment.